### PR TITLE
Support manual review

### PR DESCRIPTION
**Pull request by mitruly**
Add functionality required to support API manual review features. These features include:
- Scheduling batch changes
- Approving, rejecting, canceling batch requests that are pending review
- Supporting new request parameters (eg. `ListBatchChangesRequest`, etc.)

-----
### Note to reviewers:
Successfully tested each method against a local instance of VinylDNS:
```
VinylDNSClient client =
    new VinylDNSClientImpl(
        new VinylDNSClientConfig(
            "http://localhost:9000",
            new BasicAWSCredentials("testUserAccessKey", "testUserSecretKey")));

List<ChangeInput> changes = new ArrayList<>();
AData adata = new AData("1.2.3.4");
AddChangeInput addInput = new AddChangeInput("needs-review.dummy.", RecordType.A, 300L, adata);
changes.add(addInput);

CreateBatchRequest batchRequest = new CreateBatchRequest(changes);
batchRequest.setOwnerGroupId("d64c5fa6-33c7-4d79-8dfd-6a490cbc3a06");

batchRequest.setScheduledTime(Instant.now().plus(5, ChronoUnit.SECONDS));

VinylDNSResponse<BatchResponse> createResponse = client.createBatchChanges(batchRequest);

String reviewComment = "some rejection comment";

VinylDNSResponse<BatchResponse> res = client.rejectBatchChanges(createResponse.getValue().getId(), reviewComment);
```

### CODE REVIEW

File: src/main/java/io/vinyldns/java/model/batch/BatchChangeApprovalStatus.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.vinyldns.java.model.batch;
+
+public enum BatchChangeApprovalStatus {
+  AutoApproved,
+  Cancelled,
```
Comment (pauljamescleary) [inline src/main/java/io/vinyldns/java/model/batch/BatchChangeApprovalStatus.java]: Shouldn't this be `Canceled`? 

;P

File: src/main/java/io/vinyldns/java/model/batch/CreateBatchRequest.java [hunk 2]
```diff
@@ -57,6 +64,39 @@ public void setOwnerGroupId(String ownerGroupId) {
...
+  public Boolean getAllowManualReview() {
+    return allowManualReview;
+  }
+
+  public void setAllowManualReview(Boolean allowManualReview) {
+    this.allowManualReview = allowManualReview;
+  }
+
+  /** Returns String representation of timestamp in ISO 8601 format */
+  public String getScheduledTime() {
```
Comment (pauljamescleary) [inline src/main/java/io/vinyldns/java/model/batch/CreateBatchRequest.java]: Oh this isn't swell is it.  I presume this is consistent?  This really should be an Instant in UTC

File: src/main/java/io/vinyldns/java/VinylDNSClientImpl.java [hunk 3]
```diff
@@ -335,11 +344,40 @@ public VinylDNSClientImpl(VinylDNSClientConfig config) {
...
+
+    if (request.getAllowManualReview() != null) {
+      vinylDNSRequest.addParameter("startFrom", request.getAllowManualReview().toString());
+    }
+
+    return executeRequest(vinylDNSRequest, BatchResponse.class);
+  }
+
+  @Override
+  public VinylDNSResponse<BatchResponse> approveBatchChanges(ApproveBatchRequest request) {
```
[the remark this answers is not in the document]
Comment (mitruly) [reply src/main/java/io/vinyldns/java/VinylDNSClientImpl.java]: @pauljamescleary We could, though the convenience with a `class` is that we're using `gson` to serialize into JSON.

The alternative to this is that we would have to construct the request payload in the implementing function.

File: src/main/java/io/vinyldns/java/model/batch/BatchChangeApprovalStatus.java (hunk 1 above)
Comment (mitruly) [reply src/main/java/io/vinyldns/java/model/batch/BatchChangeApprovalStatus.java]: xD

File: src/main/java/io/vinyldns/java/model/batch/CreateBatchRequest.java (hunk 2 above)
Comment (mitruly) [reply src/main/java/io/vinyldns/java/model/batch/CreateBatchRequest.java]: Yeah, that was a bad move; I overhauled this so it should be good now.

File: src/main/java/io/vinyldns/java/VinylDNSClientImpl.java [hunk 4]
```diff
@@ -335,11 +344,56 @@ public VinylDNSClientImpl(VinylDNSClientConfig config) {
   @Override
   public VinylDNSResponse<BatchResponse> createBatchChanges(CreateBatchRequest request) {
     String path = "zones/batchrecordchanges";
+
+    VinylDNSRequest<CreateBatchRequest> vinylDNSRequest =
+        new VinylDNSRequest<>(Methods.POST.name(), getBaseUrl(), path, request);
+
+    if (request.getAllowManualReview() != null) {
+      vinylDNSRequest.addParameter("startFrom", request.getAllowManualReview().toString());
```
Comment (britneywright) [inline src/main/java/io/vinyldns/java/VinylDNSClientImpl.java]: should be `allowManualReview` not `startFrom`

File: src/main/java/io/vinyldns/java/model/batch/SingleChangeStatus.java [hunk 5]
```diff
@@ -14,7 +14,9 @@
 package io.vinyldns.java.model.batch;
 
 public enum SingleChangeStatus {
-  Pending,
   Complete,
-  Failed
+  Failed,
+  NeedsReview,
+  Pending,
+  Rejected
```
Comment (britneywright) [inline src/main/java/io/vinyldns/java/model/batch/SingleChangeStatus.java]: There should be a `Cancelled` status here as well

File: src/main/java/io/vinyldns/java/VinylDNSClient.java [hunk 6]
```diff
@@ -287,12 +287,75 @@
    * A delete and add of the same record will be treated as an update on that record set. Regardless
    * of the input order in the batch change, all deletes for the same recordset will be logically
    * applied before the adds. Current supported record types for creating a batch change are: A,
-   * AAAA, CNAME, and PTR. A batch must contain at least one change and no more than 20 changes.
+   * AAAA, CNAME, and PTR. A batch must contain at least one change and no more than the configured
```
Comment (britneywright) [inline src/main/java/io/vinyldns/java/VinylDNSClient.java]: MX and TXT are supported now too

File: src/main/java/io/vinyldns/java/VinylDNSClientImpl.java (hunk 4 above)
Comment (mitruly) [reply src/main/java/io/vinyldns/java/VinylDNSClientImpl.java]: Oof, nice catch

File: src/test/java/io/vinyldns/java/VinylDNSClientTest.java [hunk 7]
```diff
@@ -1179,6 +1182,234 @@ public void createBatchChangeSuccess() {
...
+    batchResponse.setStatus(BatchChangeStatus.Cancelled);
+    batchResponse.setApprovalStatus(BatchChangeApprovalStatus.Cancelled);
+    batchResponse.setOwnerGroupId("someOwnerGroupId");
+    batchResponse.setReviewerId("reviewerId");
+    batchResponse.setReviewComment("Looks good!");
+    batchResponse.setReviewTimestamp(new Date());
+
+    String response = client.gson.toJson(batchResponse);
+
+    wireMockServer.stubFor(
```
Comment (pauljamescleary) [inline src/test/java/io/vinyldns/java/VinylDNSClientTest.java]: hmm, I see a stub for a cancel but I don't see the call being made, amirite?

File: src/main/java/io/vinyldns/java/serializers/InstantSerializer.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import java.lang.reflect.Type;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+public class InstantSerializer implements JsonSerializer<Instant>, JsonDeserializer<Instant> {
```
Comment (pauljamescleary) [inline src/main/java/io/vinyldns/java/serializers/InstantSerializer.java]: Doesn't Google gson support Instant out of the box?

File: src/test/java/io/vinyldns/java/VinylDNSClientTest.java (hunk 7 above)
Comment (mitruly) [reply src/test/java/io/vinyldns/java/VinylDNSClientTest.java]: Yes, that is correct.

File: src/main/java/io/vinyldns/java/serializers/InstantSerializer.java (hunk 8 above)
Comment (mitruly) [reply src/main/java/io/vinyldns/java/serializers/InstantSerializer.java]: Unfortunately not for serialization; GSON serializes `java.time.Instant` as the data structure (I believe `nanos ` and `epochSeconds`). We also had to write a custom serializer for `org.joda.time.DateTime` as well.

Comment (pauljamescleary) [reply src/main/java/io/vinyldns/java/serializers/InstantSerializer.java]: boooooooooo

**Pull request merged**